src/twisted/trial/_asynctest.py

```diff
@@ -190,22 +190,24 @@ def _ebDeferTearDown(self, failure, result):
             result.stop()
         self._passed = False
 
+    @defer.inlineCallbacks
     def deferRunCleanups(self, ignored, result):
         """
         Run any scheduled cleanups and report errors (if any to the result
         object.
         """
-        d = self._runCleanups()
-        d.addCallback(self._cbDeferRunCleanups, result)
-        return d
+        failures = []
+        for f, args, kwargs in list(self._cleanups):
+            try:
+                yield f(*args, **kwargs)
+            except Exception:
+                failures.append(failure.Failure())
 
-    def _cbDeferRunCleanups(self, cleanupResults, result):
-        for flag, testFailure in cleanupResults:
-            if flag == defer.FAILURE:
-                result.addError(self, testFailure)
-                if testFailure.check(KeyboardInterrupt):
-                    result.stop()
-                self._passed = False
+        for failure_ in failures:
+            result.addError(self, failure)
+            if failure_.check(KeyboardInterrupt):
```

Comment:
should KeyboardInterrupt abort cleanups?

Reply:
It should not abort the cleanups, but it should abort _after_ running the cleanups, which is what happens here since this code runs after collecting all the cleanups results.

Reply:
I think it should be more like this:

https://github.com/graingert/twisted/blob/1bfb3b5d6966afe43e46cf38cd10f41f46096eb9/src/twisted/trial/_synctest.py#L1412-L1419